server/src/ast/parser/mod.rs

```diff
@@ -0,0 +1,375 @@
+use core::convert::identity;
+use core::str::Chars;
+use std::str::FromStr;
+
+use bigdecimal::BigDecimal;
+use parser_combinator::either::Either;
+use parser_combinator::either::Either3;
+use parser_combinator::pair::Pair;
+use parser_combinator::parser::Parser;
+use parser_combinator::parser::{match_anything, match_character, match_literal};
+use parser_combinator::repeated::RepeatedParser;
+use parser_combinator::triple::Triple;
+use parser_combinator::*;
+
+use error::ErrorMessage;
+
+use crate::ast::Localization;
+use crate::ast::LocalizedSyntaxNode;
+use crate::ast::SyntaxTree;
+use crate::most_important_of;
+
+mod error;
+#[cfg(test)]
+mod tests;
+
+type ParseResult<'a> =
+    parser_combinator::ParseResult<'a, CharWrapper<'a>, LocalizedSyntaxNode, ErrorMessage>;
+
+#[derive(Clone, Debug)]
+pub struct CharWrapper<'a> {
+    chars: Chars<'a>,
+    start: Localization,
+    end: Localization,
+}
+
+impl<'a> CharWrapper<'a> {
+    fn new(chars: Chars<'a>, start: Localization, end: Localization) -> CharWrapper<'a> {
+        Self { chars, start, end }
+    }
+}
+
+impl<'a> Iterator for CharWrapper<'a> {
+    type Item = char;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.start = self.end;
+        self.end.column += 1;
+        let next = self.chars.next();
+
+        if next == Some('\n') {
+            self.end.line += 1;
+            self.end.column = 0;
+        }
+        next
+    }
+}
+
+fn parse(input: String) -> Result<Vec<LocalizedSyntaxNode>, ErrorMessage> {
+    let (result, leftover) = parse_expression
+        .separated_by(match_character(';'))
+        .with_error(|err, _| {
+            err.fold(identity, |err2| {
+                err2.fold(ErrorMessage::forgot_comma, identity)
+            })
+        })
+        .with_error(|error, _| error.map_message(|message| format!("Syntax Error: {}", message)))
+        .transform(move |(first, rest)| {
+            let mut result = vec![first];
+            for (_, fragment) in rest {
+                result.push(fragment);
+            }
+            result
+        })
+        .skip(whitespace)
+        .skip(semicolon)
+        .parse(CharWrapper::new(
+            input.chars(),
+            Localization::new(),
+            Localization::new(),
+        ))?;
+
+    let leftover_string: String = leftover.chars.collect();
+    if leftover_string.len() != 0 {
+        return Err(ErrorMessage::leftover(leftover_string, leftover.end));
+    }
+
+    Ok(result)
+}
+
+fn parse_expression(input: CharWrapper) -> ParseResult {
+    if input.clone().filter(|char| !char.is_whitespace()).count() == 0 {
+        return Err(ErrorMessage::empty_expression(format!(
+            "expected expression, got '{}'",
+            input.chars.collect::<String>()
+        )));
+    }
+
+    let add_or_subtract = match_literal(CharWrapper::new("+".chars(), input.start, input.end))
+        .or_else(match_literal(CharWrapper::new(
```

Comment:
I don't really like that we misuse our input here. The localization part is totally misplaced.

Shouldn't the localization maybe actually be a part of Kirby itself?

Reply:
We wouldn't have the notion of columns and lines, but simply a straight counter instead. But that would be useful anyway. The translation into lines could be done by user code then? Kirby itself should not need to know about what it actually parses, i.e. newlines don't exist there.

Reply:
that makes sense